File: frontend/src/api/services/reportes.service.js
// src/api/services/reportes.service.js
import api from '../axios';

class ReportesService {
  // Dashboard
  async getDashboard() {
    const response = await api.get('/reportes/dashboard');
    return response.data;
  }

  // Usuarios
  async getResumenUsuarios(filtros = {}) {
    const params = new URLSearchParams();
    if (filtros.fechaInicio) params.append('fechaInicio', filtros.fechaInicio);
    if (filtros.fechaFin) params.append('fechaFin', filtros.fechaFin);
    if (filtros.estado) params.append('estado', filtros.estado);
    
    const response = await api.get(`/reportes/usuarios/resumen?${params.toString()}`);
    return response.data;
  }

  async getUsuariosNuevosPorDia(fechaInicio, fechaFin) {
    const response = await api.get(`/reportes/usuarios/nuevos-por-dia?fechaInicio=${fechaInicio}&fechaFin=${fechaFin}`);
    return response.data;
  }

  async getUsuariosPorRol() {
    const response = await api.get('/reportes/usuarios/por-rol');
    return response.data;
  }

  // Productos
  async getResumenProductos(filtros = {}) {
    const params = new URLSearchParams();
    if (filtros.categoria) params.append('categoria', filtros.categoria);
    if (filtros.estado) params.append('estado', filtros.estado);
    
    const response = await api.get(`/reportes/productos/resumen?${params.toString()}`);
    return response.data;
  }

  async getProductosNuevosPorDia(fechaInicio, fechaFin) {
    const response = await api.get(`/reportes/productos/nuevos-por-dia?fechaInicio=${fechaInicio}&fechaFin=${fechaFin}`);
    return response.data;
  }

  async getProductosPorCategoria() {
    const response = await api.get('/reportes/productos/por-categoria');
    return response.data;
  }

  async getEstadoStock() {
    const response = await api.get('/reportes/productos/estado-stock');
    return response.data;
  }

  // Stock
  async getResumenStock() {
    const response = await api.get('/reportes/stock/resumen');
    return response.data;
  }

  async getMovimientos(filtros = {}) {
    const params = new URLSearchParams();
    if (filtros.fechaInicio) params.append('fechaInicio', filtros.fechaInicio);
    if (filtros.fechaFin) params.append('fechaFin', filtros.fechaFin);
    if (filtros.tipo) params.append('tipo', filtros.tipo);
    
    const response = await api.get(`/reportes/stock/movimientos?${params.toString()}`);
    return response.data;
  }

  // ✅ Cotizaciones - CON TIMESTAMP PARA EVITAR CACHÉ
  async getResumenCotizaciones(filtros = {}) {
    const params = new URLSearchParams();
    if (filtros.fechaInicio) params.append('fechaInicio', filtros.fechaInicio);
    if (filtros.fechaFin) params.append('fechaFin', filtros.fechaFin);
    if (filtros.estado) params.append('estado', filtros.estado);
    
    // ✅ Agregar timestamp para evitar caché
    params.append('_t', Date.now());
    
    console.log('📊 Parámetros enviados:', params.toString());
    
    const response = await api.get(`/reportes/cotizaciones/resumen?${params.toString()}`);
    
    console.log('📊 Respuesta del servidor:', response.data);
    console.log('📊 Cantidad de cotizaciones:', response.data?.cotizaciones?.length);
    
    return response.data;
  }

  async getCotizacionesPorMetodo() {
    const response = await api.get('/reportes/cotizaciones/por-metodo');
    return response.data;
  }

  // Exportaciones
  async exportarExcel() {
    try {
      const response = await api.get('/reportes/exportar/excel', {
        responseType: 'blob',
      });
      
      const url = window.URL.createObjectURL(new Blob([response.data]));
      const link = document.createElement('a');
      link.href = url;
      link.setAttribute('download', `reporte_${new Date().toISOString().split('T')[0]}.xlsx`);
      document.body.appendChild(link);
      link.click();
      link.remove();
      window.URL.revokeObjectURL(url);
    } catch (error) {
      console.error('Error exportando Excel:', error);
      throw error;
    }
  }

  async exportarPDF() {
    try {
      const response = await api.get('/reportes/exportar/pdf', {
        responseType: 'blob',
      });
      
      const url = window.URL.createObjectURL(new Blob([response.data]));
      const link = document.createElement('a');
      link.href = url;
      link.setAttribute('download', `reporte_${new Date().toISOString().split('T')[0]}.pdf`);
      document.body.appendChild(link);
      link.click();
      link.remove();
      window.URL.revokeObjectURL(url);
    } catch (error) {
      console.error('Error exportando PDF:', error);
      throw error;
    }
  }
}

export default new ReportesService();

File: lib/core/network/reportes_service.dart
// lib/core/network/reportes_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:gramas_y_suministros_movil/core/network/api_config.dart';

class ReportesService {
  final http.Client _client = http.Client();

  // Dashboard
  Future<Map<String, dynamic>> getDashboard(String token) async {
    try {
      final response = await _client.get(
        Uri.parse('${ApiConfig.reportes}/dashboard'),
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        Map<String, dynamic> data = jsonDecode(response.body);
        
        if (!data.containsKey('stockDetalle') || data['stockDetalle'] == null) {
          try {
            final stockResponse = await _client.get(
              Uri.parse('${ApiConfig.reportes}/productos/estado-stock'),
              headers: {
                'Authorization': 'Bearer $token',
                'Content-Type': 'application/json',
              },
            );
            if (stockResponse.statusCode == 200) {
              final stockData = jsonDecode(stockResponse.body);
              data['stockDetalle'] = stockData['detalle'] ?? [];
            }
          } catch (e) {
            data['stockDetalle'] = [];
          }
        }
        
        return data;
      } else {
        throw Exception('Error al cargar dashboard: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error de conexión: $e');
    }
  }

  // Usuarios
  Future<Map<String, dynamic>> getResumenUsuarios({
    required String token,
    String? fechaInicio,
    String? fechaFin,
    String? estado,
  }) async {
    try {
      final queryParams = <String, String>{};
      if (fechaInicio != null) queryParams['fechaInicio'] = fechaInicio;
      if (fechaFin != null) queryParams['fechaFin'] = fechaFin;
      if (estado != null) queryParams['estado'] = estado;

      final uri = Uri.parse('${ApiConfig.reportes}/usuarios/resumen')
          .replace(queryParameters: queryParams);

      final response = await _client.get(
        uri,
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('Error al cargar usuarios: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error de conexión: $e');
    }
  }

  // Productos
  Future<Map<String, dynamic>> getResumenProductos({
    required String token,
    String? categoria,
    String? estado,
  }) async {
    try {
      final queryParams = <String, String>{};
      if (categoria != null) queryParams['categoria'] = categoria;
      if (estado != null) queryParams['estado'] = estado;

      final uri = Uri.parse('${ApiConfig.reportes}/productos/resumen')
          .replace(queryParameters: queryParams);

      final response = await _client.get(
        uri,
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('Error al cargar productos: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error de conexión: $e');
    }
  }

  // Stock
  Future<Map<String, dynamic>> getEstadoStock(String token) async {
    try {
      final response = await _client.get(
        Uri.parse('${ApiConfig.reportes}/productos/estado-stock'),
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('Error al cargar stock: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error de conexión: $e');
    }
  }

  // Cotizaciones
  Future<Map<String, dynamic>> getResumenCotizaciones({
    required String token,
    String? fechaInicio,
    String? fechaFin,
    String? estado,
  }) async {
    try {
      final queryParams = <String, String>{};
      if (fechaInicio != null) queryParams['fechaInicio'] = fechaInicio;
      if (fechaFin != null) queryParams['fechaFin'] = fechaFin;
      if (estado != null) queryParams['estado'] = estado;

      final uri = Uri.parse('${ApiConfig.reportes}/cotizaciones/resumen')
          .replace(queryParameters: queryParams);

      final response = await _client.get(
        uri,
        headers: {
          'Authorization': 'Bearer $token',
          'Content-Type': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        throw Exception('Error al cargar cotizaciones: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error de conexión: $e');
    }
  }

  // Exportar Excel
  Future<http.Response> exportarExcel(String token) async {
    try {
      final response = await _client.get(
        Uri.parse('${ApiConfig.reportes}/exportar/excel'),
        headers: {
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        return response;
      } else {
        throw Exception('Error al exportar Excel: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error de conexión: $e');
    }
  }

  // Exportar PDF
  Future<http.Response> exportarPDF(String token) async {
    try {
      final response = await _client.get(
        Uri.parse('${ApiConfig.reportes}/exportar/pdf'),
        headers: {
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        return response;
      } else {
        throw Exception('Error al exportar PDF: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error de conexión: $e');
    }
  }
}